### ISSUE TITLE
[BUG] Possible issue with DDP and asr decoder gradient syncing

### ISSUE THREAD

**Issue by janvainer** [k2-fsa/icefall#226]
Hi, I took some inspiration from the training script for conformer DDP training. There were some convergence issues for my usecase on 4 GPUs so I digged a bit and noticed that model weights on each GPU are different. It was because I used something like
```
model = DDP(model, ...)
...

# Training
encoder_out = model.module.forward_encoder()
decoder_out = model.module.forward_decoder()
loss = loss_fn(decoder_out, targets)
loss.backwar()
...
```
The issue seems to be that calling `model.module.forward_sth` on the DDP wrapper effectively avoids the grad synchronization for weights affected by the function call. (I guess DDP only puts hooks on the `forward` method).

The asr decoder in icefall is doing the same - see this [block of code](https://github.com/k2-fsa/icefall/blob/ac7c2d84bc44dd83a5653f68677ae1ef16551eea/egs/librispeech/ASR/conformer_ctc/train.py#L404).
Its possible that your ASR decoders are different on each GPU and you are training  `world_size different` versions of them.
I am not sure if its intentional (maybe it regularizes something) or if it should be fixed. Anyways, just letting you know about this :)

Relevant pytorch forum thread [here](https://discuss.pytorch.org/t/model-in-distributeddataparallel-must-implement-and-call-forward-funciton/142807).

**Comment by danpovey** [comment]
Was your example calling code ever calling `model(inputs)` (i.e. the conventional forward function)?
I believe if you first call the regular forward and later call other-named functions, it's all OK.

**Comment by csukuangfj** [comment]
I just wrote a demo to verify that the code in icefall does work.

```python3
#!/usr/bin/env python3

import os
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.nn.parallel import DistributedDataParallel as DDP

import torch
import torch.nn as nn

import datetime


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear0 = nn.Linear(1, 1)
        self.linear1 = nn.Linear(1, 1)
        self.linear2 = nn.Linear(1, 1)

    def forward(self, x: torch.Tensor):
        return self.linear0(x)

    def forward1(self, x: torch.Tensor):
        return self.linear1(x)

    def forward2(self, x: torch.Tensor):
        return self.linear2(x)

def run(rank: int, world_size: int):
    print(f"world_size: {world_size}")
    device = torch.device("cuda", rank)

    model = Model()
    model.to(device)
    model = DDP(model, device_ids=[rank])
    print("before update", "rank", rank, list(model.named_parameters()))

    x0 = torch.tensor([rank], dtype=torch.float32, requires_grad=True, device=device)
    x1 = torch.tensor(
        [rank + 100], dtype=torch.float32, requires_grad=True, device=device
    )
    x2 = torch.tensor(
        [rank + 1000], dtype=torch.float32, requires_grad=True, device=device
    )

    y0 = model(x0)
    y1 = model.module.forward1(x1)
    y2 = model.module.forward2(x2)

    y = y0 + y1 + y2

    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
    optimizer.zero_grad()
    y.backward()
    optimizer.step()
    print("after update", "rank", rank, list(model.named_parameters()))

def init_process(rank: int, world_size: int, fn):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12357"
    dist.init_process_group(
        "nccl", rank=rank, world_size=world_size, timeout=datetime.timedelta(0, 5)
    )
    fn(rank, world_size)


if __name__ == "__main__":
    print(f"dist.is_available: {dist.is_available()}")
    world_size = 3
    processes = []
    mp.set_start_method("spawn")
    for rank in range(world_size):
        p = mp.Process(target=init_process, args=(rank, world_size, run))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

The output is given below and it shows that the parameters for models on different devices are the same.

```
dist.is_available: True
world_size: 3
world_size: 3
world_size: 3
before update rank 0 [('module.linear0.weight', Parameter containing:
tensor([[-0.2952]], device='cuda:0', requires_grad=True)), ('module.linear0.bias', Parameter containing:
tensor([0.1852], device='cuda:0', requires_grad=True)), ('module.linear1.weight', Parameter containing:
tensor([[0.6371]], device='cuda:0', requires_grad=True)), ('module.linear1.bias', Parameter containing:
tensor([-0.5620], device='cuda:0', requires_grad=True)), ('module.linear2.weight', Parameter containing:
tensor([[0.5978]], device='cuda:0', requires_grad=True)), ('module.linear2.bias', Parameter containing:
tensor([-0.6750], device='cuda:0', requires_grad=True))]
before update rank 2 [('module.linear0.weight', Parameter containing:
tensor([[-0.2952]], device='cuda:2', requires_grad=True)), ('module.linear0.bias', Parameter containing:
tensor([0.1852], device='cuda:2', requires_grad=True)), ('module.linear1.weight', Parameter containing:
tensor([[0.6371]], device='cuda:2', requires_grad=True)), ('module.linear1.bias', Parameter containing:
tensor([-0.5620], device='cuda:2', requires_grad=True)), ('module.linear2.weight', Parameter containing:
tensor([[0.5978]], device='cuda:2', requires_grad=True)), ('module.linear2.bias', Parameter containing:
tensor([-0.6750], device='cuda:2', requires_grad=True))]
before update rank 1 [('module.linear0.weight', Parameter containing:
tensor([[-0.2952]], device='cuda:1', requires_grad=True)), ('module.linear0.bias', Parameter containing:
tensor([0.1852], device='cuda:1', requires_grad=True)), ('module.linear1.weight', Parameter containing:
tensor([[0.6371]], device='cuda:1', requires_grad=True)), ('module.linear1.bias', Parameter containing:
tensor([-0.5620], device='cuda:1', requires_grad=True)), ('module.linear2.weight', Parameter containing:
tensor([[0.5978]], device='cuda:1', requires_grad=True)), ('module.linear2.bias', Parameter containing:
tensor([-0.6750], device='cuda:1', requires_grad=True))]
after update rank 2 [('module.linear0.weight', Parameter containing:
tensor([[-0.2962]], device='cuda:2', requires_grad=True)), ('module.linear0.bias', Parameter containing:
tensor([0.1842], device='cuda:2', requires_grad=True)), ('module.linear1.weight', Parameter containing:
tensor([[0.6361]], device='cuda:2', requires_grad=True)), ('module.linear1.bias', Parameter containing:
tensor([-0.5630], device='cuda:2', requires_grad=True)), ('module.linear2.weight', Parameter containing:
tensor([[0.5968]], device='cuda:2', requires_grad=True)), ('module.linear2.bias', Parameter containing:
tensor([-0.6760], device='cuda:2', requires_grad=True))]
after update rank 0 [('module.linear0.weight', Parameter containing:
tensor([[-0.2962]], device='cuda:0', requires_grad=True)), ('module.linear0.bias', Parameter containing:
tensor([0.1842], device='cuda:0', requires_grad=True)), ('module.linear1.weight', Parameter containing:
tensor([[0.6361]], device='cuda:0', requires_grad=True)), ('module.linear1.bias', Parameter containing:
tensor([-0.5630], device='cuda:0', requires_grad=True)), ('module.linear2.weight', Parameter containing:
tensor([[0.5968]], device='cuda:0', requires_grad=True)), ('module.linear2.bias', Parameter containing:
tensor([-0.6760], device='cuda:0', requires_grad=True))]
after update rank 1 [('module.linear0.weight', Parameter containing:
tensor([[-0.2962]], device='cuda:1', requires_grad=True)), ('module.linear0.bias', Parameter containing:
tensor([0.1842], device='cuda:1', requires_grad=True)), ('module.linear1.weight', Parameter containing:
tensor([[0.6361]], device='cuda:1', requires_grad=True)), ('module.linear1.bias', Parameter containing:
tensor([-0.5630], device='cuda:1', requires_grad=True)), ('module.linear2.weight', Parameter containing:
tensor([[0.5968]], device='cuda:1', requires_grad=True)), ('module.linear2.bias', Parameter containing:
tensor([-0.6760], device='cuda:1', requires_grad=True))]
```

**Comment by csukuangfj** [comment]
> I believe if you first call the regular forward and later call other-named functions, it's all OK.

I just tested with the above code and find that it is not required to call `forward()` to make DDP work.
```python3
y = model.module.forward2(x2)
y.backward()
```
also works in the above demo.

**Comment by danpovey** [comment]
OK, I'm a bit confused now, because I thought the forward() of the DDP module is when it sets up the buckets for DDP.  I'm surprised if it is working without calling that.

**Comment by danpovey** [comment]
... perhaps the issue is that with Adam update, the update only depends on the sign of the grad (for the 1st step).

**Comment by csukuangfj** [comment]
https://github.com/pytorch/pytorch/blob/186ef8b22d229c80be8f1c959476d84eeb3e8b91/torch/nn/parallel/distributed.py#L931
shows the implementation of the `forward()` method of a DDP model. In addition to call `model.module.forward()`, it does extra stuff, though I am not clear how important is that.

**Comment by janvainer** [comment]
Thanks for the answers! I also tried to set torch print option to more decimal points and your code really works fine. It also works with Lamb optimizer. It also works fine for multiple steps. I have no idea what was the issue in my code then. Anyways, thank you for looking into it!
